text/0496-handlebars-strict-mode.md

```diff
@@ -0,0 +1,591 @@
+import BlogPost from './components/blog-post';
+---
+
+{{#let this.session.currentUser as |user|}}
+  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+{{/let}}
+```
+
+The build tool can compile this into a JavaScript module like this:
+
+```js
+import { createTemplateFactory } from '@ember/template-factory';
+import { titleize } from '@ember/template-helpers';
+import BlogPost from './components/blog-post';
+
+export default createTemplateFactory({
+  "id": "ANJ73B7b",
+  "block": "{\"statements\":[\"...\"]}",
+  "meta": { "moduleName": "index.hbs" },
+  "scope": () => [BlogPost, titleize]
+});
+```
+
+When this is evaulated by a JavaScript engine, the references in the `scope`
+closure will automatically be "linked up" with the imports, and Ember will be
+able to reference these values when rendering the template. Note that these
+references are _static_–the values are essentially "snapshotted" by the
+rendering engine whenever the template is instantiated. Updates to these values
+in the JavaScript scope will _not_ be observable by the rendering engine, even
+in conjunction with `Ember.set` or `@tracked`.
+
+Optionally, the build tool can choose to restrict the set of allowed ambient
+references by suppling an array of available identifiers to the compiler:
+
+```js
+import { precompileTemplate } from '@ember/template-compilation';
+
+precompileTemplate(`{{#let this.session.currentUser as |user|}}
+  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+{{/let}}`, {
+  moduleName: 'index.hbs',
+  strict: true,
+  scope: ['BlogPost', 'titleize']
+});
+```
+
+If the template compiler encounters any undefined references outside of this
+allowed list, it will throw an error with the appropiate location info. It also
+follows that build tools can choose to disable this feature completely by
+passing an empty array.
+
+### Keywords
+
+While most items should be imported into scope explicitly, some of the existing
+constructs in the language are unimportable will be made available as keywords
+instead:
+
+* `action`
+* `debugger`
+* `each-in`
+* `each`
+* `has-block-params`
+* `has-block`
+* `hasBlock`
+* `if`
+* `in-element`
+* `let`
+* `link-to` (non-block form curly invocations)
+* `loc`
+* `log`
+* `mount`
+* `mut`
+* `outlet`
+* `query-params`
+* `readonly`
+* `unbound`
+* `unless`
+* `with`
+* `yield`
+
+These keywords do _not_ have to be imported into scope and will always be
+ambiently available.
+
+On the other hand, the following built-in constructs will need to be imported
+(the current or proposed import paths in parentheses):
+
+* `array` (`import { array } from '@ember/helper`)
+* `concat` (`import { concat } from '@ember/helper`)
+* `fn` (`import { fn } from '@ember/helper`)
+* `get` (`import { get } from '@ember/helper`)
+* `hash` (`import { hash } from '@ember/helper`)
+* `on` (`import { on } from '@ember/modifier'`)
```

Comment:
after working with gjs/gts for some time now, I find importing some of these to be quite cumbersome -- in particular `on` and `fn`, as they are my most used tools from this list.

Why not make a subset of this list built-in / non-importable?
on and fn are _essential_ features to getting anything done in templates and they feel like great candidates for being "always available"

Reply:
Totally agree!

Reply:
Regarding `fn`, I would prefer to see a syntax proposal for embedding javascript expression (not statements!). This would naturally cover the `fn` case because an arrow function expression would be legal there. It would also replace the need to import the `hash` and `array` helpers.

Reply:
I don't know if this is the case at the moment, but if typing `{{on` + enter would autocomplete the import statement, that would alleviate the problem I think?

Reply:
not really. 
and tooling still greatly varies between editors and size of project. on big monorepos, ts can take almost a minute to boot. fol demos, markdown examples, repls, etc, we can't rely on tooling either.

additionally, it's visual noise at the top of the file.

Reply:
Personally, I would keep it as is, I also don't see any difference between having to import `on` or having to import for example `service` from `@ember/service`, which is used a lot as well.

Reply:
I don't have the ability to grab stats anymore, but i feel like they are not the same. On/fn are used way more than service - in forms, modals, many many places more frequent than where service is needed.

Even so, do we not want to make the framework easier to use and less cumbersome?

On and fn feel like language level features, which are a whole class of their own, compared to service.